### ISSUE TITLE
REST API

### ISSUE THREAD

**Issue by grantcarthew** [grantcarthew/rjq-ui-design#4]
At some point there may be a need to add new methods into `rethinkdb-job-queue` to expose reporting information to support the UI frontend.

The REST API will expose most of this functionality. If we need to extend the Queue API to support it then we will.

**Comment by marcoalbarelli** [comment]
As stated in #3 I'd be glad to lend a helping hand.
Before starting with the actual code I need to understand the actual architecture we're going to implement.
I read #1 but I still have some doubts

Let me try to summarize the architecture so far (and correct me where I'm wrong): 
A RJQ-UI instance is running somewhere and it's connected to the db cluster where the queue is hosted too
The RJQ-UI has its own database somewhere (configured via env variables)
The queue names and other values are configured via DB entries
The RJQ-UI backend provides APIs for the frontend, performing all needed actions

I take it we're in the process of defining/mocking the UI APIs. How is that task going?
Scaffolding a standard express app takes little time and gives little value :)
It would be nice to implement the mocks at least
Are the ones in the Backend.md file valid?

**Comment by sagivf** [comment]
@marcoalbarelli I don't think RJQ-UI has it's own UI.
It needs to be connected to the cluster.
It can get the needed info regarding Jobs and Queues in two ways:
1) Directly querying the database
2) Create an Queue Object and query that
It can be one of them or both.
This is what I mean by bootstrapping the server, start figuring out which option and what is available.
The mocks are currently defined here:
https://github.com/sagivf/rjq-ui/blob/master/src/mock.js

**Comment by marcoalbarelli** [comment]
Ok, then I think we're on the same page

The reason behind providing RJQ-UI its own db is that I'd like to have at least some rudimental authentication system starting from day 0
It could be an optional requirement, having a fallback on env provided credentials if no db connection is configured.

**Comment by sagivf** [comment]
Sounds reasonable, don't forget Rethink It self has a number of security options.
My efforts are currently on the frontend side of things, working about 1 day a week give or take.
Would be more then happy to see contribution :)

**Comment by marcoalbarelli** [comment]
Yeah, I'll have to dive a little into Rethinkdb auth features, even though my _real_ goal is to have a simple ACL-like functionality and I doubt rethinkdb's auth features will be enough for that, but some RTFM is mandatory :)

**Comment by grantcarthew** [comment]
After the discussions in these issues I see the overall project as three integral parts:

1.   `rethinkdb-job-queue` - Job publishing and working
2.   `rjq-api` - Backend API     <= Name and project not fixed yet
3.   `rjq-ui` - Web front-end   <= Sagiv has started working on this

Following is just my own ideas guys. Purely food for thought. I am suppose to be studying MS SQL Server right now... 😳

## `rjq-api`

This project will run expressjs and expose a REST API to return queue data in JSON format for consumption by any project. This would allow someone for example to publish jobs to the queue by using just an nginx server or any capable client.

From an authentication perspective you could go with a configuration file created with a CLI interface. Run the rjq-api command line tool and it asks for your username/password/role and saves the password in a hash format to the config file. Security here is based on the server access. If someone has access to your server you don't have security anyway.

The other option for security is to use the RethinkDB security options. This would involve adding security to the Queue object or passing in the `rethinkdbdash` driver authenticated. I'm speculating here as I have not used the RethinkDB user security features yet.

## `rjq-ui`

This [project already underway](https://github.com/sagivf/rjq-ui) will be a web front-end that could be served by a standard web server like nginx or could instantiate an expressjs server by calling a method in code. Something like this:

```js

const rjqUi = require('rjq-ui')
const rjqUiServer = new rjqUiFactory({ //connection options such as IP:Port etc })

``` 

By having expressjs in the UI package it makes it easy for someone to serve the UI on their queue server with little fuss.

**Comment by grantcarthew** [comment]
I created a Slack channel under the RethinkDB for real time discussion:

https://rethinkdb.slack.com/messages/rethinkdb-job-queue/

Join in if you wish.

**Comment by grantcarthew** [comment]
This is of interest: https://github.com/rethinkdb/rethinkdb/issues/1163

**Comment by sagivf** [comment]
I like the idea of `rjq-api` being initiated via the web ui.

Would be even better if:
1) You could add a configuration that would point the static resources to the resources created by webpack. Connecting it to an ENV_VARIABLE would even make it `heroku`/other ready.
2) Configure via webpack the server location in case you want the api not to sit along side the UI.

**Comment by marcoalbarelli** [comment]
`
Don't have an account on this team yet?
Contact the team administrator for an invitation
` :)

Regarding rethinkdb auth, I think it's a no-go as with all other dbs
The `save your user/roles in a file via a smart little helper` approach is fine for me: I'll just need to read them from FS when authenticating the user in passport
This approach is good enough for orientDB and will postpone the need for a dedicated db, making deployment a little easier.
